### PR TITLE
Fix StackOverflowError coming from `MOI.Utilities.operate`

### PR DESCRIPTION
**Pull request by bachdavi**

### CODE REVIEW

File: src/SolverAPI.jl [hunk 1]
```diff
@@ -469,8 +470,20 @@ function nl_to_aff_or_quad(::Type{T}, f::MOI.ScalarNonlinearFunction) where {T<:
...
         else
-            h = get(_quad_ops, f.head, nothing)
-            isnothing(h) || return MOI.Utilities.operate(h, T, args...)
+            if f.head == :+
+                # TODO (dba) this is a workaround to avoid a
+                # `StackOverflowError` error coming from
+                # `MOI.Utilities.operate`. We should avoid the `...`
+                # all together.
+                if args isa Vector{MOI.ScalarAffineFunction{Float64}}
+                    return MA.operate!(+, args[1], args[2...])
```
Comment (hdavid16) [inline src/SolverAPI.jl:479]: Did you intend to have the splatting outside the brackets? `args[2]...`

File: test/solve_tests.jl [hunk 2]
```diff
@@ -9,9 +9,10 @@ import MathOptInterface as MOI
 export run_solve, read_json
 
 function _get_solver(solver_name::String)
-    if solver_name == "MiniZinc"
+    solver_name_lower = lowercase(solver_name)
```
Comment (hdavid16) [inline test/solve_tests.jl:12]: Good idea here. Makes it more robust.

File: src/SolverAPI.jl [hunk 3]
```diff
@@ -469,8 +470,20 @@ function nl_to_aff_or_quad(::Type{T}, f::MOI.ScalarNonlinearFunction) where {T<:
                 return MOI.Utilities.operate(*, T, args[1], args[1])
             end
         else
-            h = get(_quad_ops, f.head, nothing)
-            isnothing(h) || return MOI.Utilities.operate(h, T, args...)
+            if f.head == :+
```
Comment (hdavid16) [inline src/SolverAPI.jl:473]: We should have a more general solution here. The `head` could be `:-` with 1000 terms.

File: src/SolverAPI.jl (hunk 1 above)
Comment (hdavid16) [reply src/SolverAPI.jl:479]: Also, should SAF be typed to `Float64` or to `T`?

Comment (hdavid16) [reply src/SolverAPI.jl:479]: `MA.operate` is still recursive...

File: src/SolverAPI.jl [hunk 4]
```diff
@@ -462,32 +462,74 @@ json_to_snf(a::Real, ::Dict) = a
...
+
 _quad_ops = Dict(:+ => +, :- => -, :* => *, :/ => /)
 
 function canonicalize_SNF(::Type{T}, f) where {T<:Real}
-    try
-        f = nl_to_aff_or_quad(T, f)
-    catch
-    end
-    return f
+    return nl_to_aff_or_quad(T, f)
```
Comment (bachdavi) [inline src/SolverAPI.jl:532]: This hides the underlying exception and makes it harder to debug. @chriscoey Are there cases where we can actually continue even though we can't convert the SNF?

Comment (bachdavi) [reply src/SolverAPI.jl:546]: Nevermind I found the reason :)

File: src/SolverAPI.jl (hunk 3 above)
Comment (bachdavi) [reply src/SolverAPI.jl]: It appears that `operate(-, ...)` is only defined for two arguments if I am not mistaken.

File: src/SolverAPI.jl (hunk 1 above)
Comment (bachdavi) [reply src/SolverAPI.jl]: Yeah this code was bogus ^^

**Comment by chriscoey** [comment]
[as discussed here](https://relationalai.slack.com/archives/C011C0JBE87/p1695313478255299?thread_ts=1694698939.921529&cid=C011C0JBE87) - let's attempt to fix this upstream in MOI rather than adding to the complexity of SolverAPI

File: src/SolverAPI.jl (hunk 3 above)
Comment (chriscoey) [reply src/SolverAPI.jl]: it seems to have a single argument here https://github.com/jump-dev/MathOptInterface.jl/blob/53ec2693c8280d0e3f53336142c8739298eef0e5/src/Utilities/operate.jl#L1174 for unary minus (e.g. `-5`)

**Comment by chriscoey** [comment]
The MOI fix https://github.com/jump-dev/MathOptInterface.jl/pull/2285 unblocks us for the test case added here, so lets get the test case in after MOI is tagged and leave this PR up here as DNMY. Not super high priority but when we become more concerned with performance, I think we could attempt an even larger change here that builds the SAF/SQF/SNF directly bottom-up from the JSON3 data structure, rather than going through the SNF intermediate. Because right now we do JSON->SNF with `json_to_snf` then SNF->SAF/SQF/SNF (whichever most canonical form we can convert to) with `nl_to_aff_or_quad`.

**Comment by bachdavi** [comment]
@chriscoey Sure we can wait for the upstream fix. I'm leaving this PR open in case we want to improve the performance of `nl_aff_or_quad`. I've taken the `large-model` from another one of your PRs and the stack-based version is around 3x faster and consumes 2x less memory (both running with the MOI fix). It makes a lot of sense to try to directly go from JSON to the respective function instead of this intermediate.

**Comment by chriscoey** [comment]
Thanks @bachdavi. Let's keep working on this PR to try to extend it to go directly from JSON. Hopefully that will improve the efficiency even further.

**Comment by chriscoey** [comment]
Since the [upstream MOI fix is controversial](https://github.com/jump-dev/MathOptInterface.jl/issues/2284), I decided to get this PR in rather than wait. We can further improve the performance in future PRs.

I split out the JSON-to-MOI functions into a separate file to improve clarity.

**Pull request merged**